src/EditorFeatures/Core.Wpf/InlineRename/AbstractEditorInlineRenameService.SymbolRenameInfo.cs

```diff
@@ -241,6 +238,30 @@ public bool TryOnAfterGlobalSymbolRenamed(Workspace workspace, IEnumerable<Docum
                 return _refactorNotifyServices.TryOnAfterGlobalSymbolRenamed(workspace, changedDocumentIDs, RenameSymbol,
                     this.GetFinalSymbolName(replacementText), throwOnFailure: false);
             }
+
+            public InlineRenameFileRenameInfo GetFileRenameInfo()
+            {
+                var kindSupportsRename = RenameSymbol.Kind == SymbolKind.NamedType;
+                var allowRename = kindSupportsRename 
+                    && (RenameSymbol.Locations.Length == 1)
+                    && (OriginalNameMatches(_document, RenameSymbol.Name) || ContainsOnlyOneType(_document));
+
+                return new InlineRenameFileRenameInfo(kindSupportsRename, allowRename);
+
+                // Local Functions
+
+                static bool OriginalNameMatches(Document document, string name)
+                {
+                    return document.Name.Substring(0, document.Name.Length - 3) == name;
+                }
+
+                static bool ContainsOnlyOneType(Document document)
+                {
+                    var syntaxRoot = document.GetSyntaxRootSynchronously(CancellationToken.None);
```

Comment:
UI code depending on synchronous parsing makes me sad.   might not be terrible to do this async, and pop in the checkmark when we know... (but that would add a lot of async compexity to these codepaths).